targets/stm32/mqtts/src/MQTTInterface.c

```diff
@@ -0,0 +1,262 @@
+}
+
+int net_init(Network *n)
+{
+	int ret;
+
+	// Initialize mbedTLS realted variables
+#if defined(MBEDTLS_MEMORY_BUFFER_ALLOC_C)
+	mbedtls_memory_buffer_alloc_init(alloc_buf, sizeof(alloc_buf));
+#endif
+
+#if defined(MBEDTLS_DEBUG_C)
+	mbedtls_debug_set_threshold(DEBUG_LEVEL);
+#endif
+
+	// Mbedtls_net_init(&server_fd); //MX_LWIP_Init() is called already
+	mbedtls_ssl_init(&ssl);
+	mbedtls_ssl_config_init(&conf);
+	mbedtls_x509_crt_init(&cacert);
+	mbedtls_ctr_drbg_init(&ctr_drbg);
+
+	mbedtls_entropy_init(&entropy);
+	if ((ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy, (const unsigned char *)pers,
+									 strlen(pers))) != 0)
+	{
+		return -1;
+	}
+
+	// Register functions
+	n->mqttread = netRead;
+	n->mqttwrite = netWrite;
+	n->disconnect = netDisconnect;
+
+	return 0;
+}
+
+int netConnect(Network *n, char *ip, int port)
+{
+	int ret;
+
+	// SSL/TLS connection process.
+	ret = mbedtls_x509_crt_parse(&cacert, (const unsigned char *)mbedtls_root_certificate,
+								 mbedtls_root_certificate_len);
+	if (ret < 0)
+	{
+		printf("mbedtls_x509_crt_parse failed.\n");
+		return -1;
+	}
+
+	ret = mbedtls_net_connect(&server_fd, server, SERVER_PORT,
+							  MBEDTLS_NET_PROTO_TCP);
+	if (ret < 0)
+	{
+		printf("mbedtls_net_connect failed.\n");
+		return -1;
+	}
+
+	ret = mbedtls_ssl_config_defaults(&conf, MBEDTLS_SSL_IS_CLIENT,
+									  MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT);
+	if (ret < 0)
+	{
+		printf("mbedtls_ssl_config_defaults failed.\n");
+		return -1;
+	}
+
+	mbedtls_ssl_conf_authmode(&conf, MBEDTLS_SSL_VERIFY_OPTIONAL);
+	mbedtls_ssl_conf_ca_chain(&conf, &cacert, NULL);
+	mbedtls_ssl_conf_rng(&conf, mbedtls_ctr_drbg_random, &ctr_drbg);
+	mbedtls_ssl_conf_dbg(&conf, my_debug, stdout);
+
+	ret = mbedtls_ssl_setup(&ssl, &conf);
+	if (ret < 0)
+	{
+		printf("mbedtls_ssl_setup failed.\n");
+		return -1;
+	}
+
+	ret = mbedtls_ssl_set_hostname(&ssl, SERVER_NAME);
+	if (ret < 0)
+	{
+		printf("mbedtls_ssl_set_hostname failed.\n");
+		return -1;
+	}
+
+	mbedtls_ssl_set_bio(&ssl, &server_fd, mbedtls_net_send, mbedtls_net_recv,
+						NULL);
+
+	while ((ret = mbedtls_ssl_handshake(&ssl)) != 0)
+	{
+		if (ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE)
+		{
+			printf("mbedtls_ssl_handshake failed.\n");
+			return -1;
+		}
+	}
+
+	ret = mbedtls_ssl_get_verify_result(&ssl);
+	if (ret < 0)
+	{
+		printf("mbedtls_ssl_get_verify_result failed.\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+int netRead(Network *n, unsigned char *buffer, int len, int timeout_ms)
+{
+	int ret;
+	int received = 0;
+	int error = 0;
+	int complete = 0;
+
+	// Set timeout
+	if (timeout_ms != 0)
+	{
+		mbedtls_ssl_conf_read_timeout(&conf, timeout_ms);
+	}
+
+	do
+	{
+		ret = mbedtls_ssl_read(&ssl, buffer, len);
+		if (ret > 0)
+		{
+			received += ret;
+		}
+		else if (ret != MBEDTLS_ERR_SSL_WANT_READ)
+		{
+			error = 1;
+		}
+		if (received >= len)
+		{
+			complete = 1;
+		}
+	} while (!error && !complete);

```

Comment:
Are thse used as a bools? If yes, do not declare them as ints - it is only harder to read plus takes more space.